src/Sources/Coalescence.jl

```diff
@@ -3,425 +3,524 @@
-    NProgMoms::Array{Int};
-    dist_thresholds = nothing,
-    norms = (FT(1), FT(1)),
-) where {N, M, FT <: Real}
-    Ndist = length(NProgMoms)
-    matrix_of_kernels = Array{CoalescenceTensor{N, FT, M}}(undef, Ndist, Ndist)
-    if kernel isa CoalescenceTensor
-        matrix_of_kernels .= get_normalized_kernel_tensor(kernel, norms)
-    else
-        @assert size(kernel) == (Ndist, Ndist)
-        for (k, ker) in enumerate(kernel)
-            matrix_of_kernels[k] = get_normalized_kernel_tensor(ker, norms)
+        new{N, P, FT, T}(N_mom_max, N_2d_ints, thresholds, kernels)
+    end
+
+    function CoalescenceData(
+        kernel::CoalescenceTensor{P, FT, T},
+        NProgMoms::NTuple{N, Int},
+        dist_thresholds::NTuple{N, FT},
+        norms::Tuple{FT, FT} = (FT(1), FT(1)),
+    ) where {N, P, T, FT <: Real}
+
+        kernel_ = get_normalized_kernel_tensor(kernel, norms)
+        kernels = ntuple(N) do j
+            ntuple(N) do k
+                kernel_
+            end
         end
+
+        CoalescenceData(kernels, NProgMoms, dist_thresholds, norms)
     end
 
-    Q = zeros(FT, Ndist, Ndist)
-    R = zeros(FT, Ndist, Ndist)
-    S = zeros(FT, Ndist, 2)
+end
 
-    kernels_order = [N for ker in matrix_of_kernels]
-    Nmom = maximum(NProgMoms) + maximum(kernels_order)
-    moments = zeros(FT, Ndist, Nmom)
+"""
+get_coal_ints(::AnalyticalCoalStyle, pdists::Array{ParticleDistribution{FT}}, coal_data::NamedTuple)
 
-    N_2d_ints = diag(kernels_order) .+ [i < Ndist ? max(NProgMoms[i], NProgMoms[i + 1]) : NProgMoms[i] for i in 1:Ndist]
-    finite_2d_ints = [zeros(FT, N_2d_ints[i], N_2d_ints[i]) for i in 1:Ndist]
+  - `pdists`: array of PSD subdistributions
+  - `coal_data`: coalescence data struct
+Updates the collision-coalescence integrals.
+"""
+function get_coal_ints(
+    cs::AnalyticalCoalStyle,
+    pdists::NTuple{N, PrimitiveParticleDistribution{FT}},
+    coal_data::CoalescenceData{N, P, FT},
+) where {N, P, FT <: Real}
 
-    coal_ints = zeros(FT, sum(NProgMoms))
+    NProgMoms = map(pdists) do dist
+        nparams(dist)
+    end
 
-    dist_thresholds = dist_thresholds == nothing ? ones(FT, Ndist) * Inf : dist_thresholds ./ norms[2]
-    @assert length(dist_thresholds) == Ndist
+    moments = get_moments_matrix(pdists, Val(P + 2), coal_data.N_mom_max)
+    finite_2d_ints = get_finite_2d_integrals(pdists, coal_data.dist_thresholds, moments, coal_data.N_2d_ints)
+    (; Q, R, S) = get_coalescence_integral_moment_qrs(cs, moments, NProgMoms, finite_2d_ints, coal_data.kernels)
 
-    return (
-        Q = Q,
-        R = R,
-        S = S,
-        moments = moments,
-        finite_2d_ints = finite_2d_ints,
-        coal_ints = coal_ints,
-        matrix_of_kernels = matrix_of_kernels,
-        dist_thresholds = dist_thresholds,
-    )
+    coal_ints = ntuple(N) do k
+        ntuple(NProgMoms[k]) do m
+            if k == 1
+                sum(@views Q[m][:, k]) - sum(@views R[m][:, k]) + S[m][1, k]
```

Comment:
@charleskawczynski mentioned that `@views` are likely problematic -- can we rethink this?

Reply:
Unfortunately, I don't remember Charlie's comment on views exactly. But in performance tips for Julia (https://docs.julialang.org/en/v1/manual/performance-tips/) they say it might be good to use views when doing operations on slices.

Reply:
The julia performance tips applies to when you are performing in-place operations, to avoid allocating intermediate arrays. I think we can write the first part as: `sum(ξ->Q[m][ξ, k], 1:size(Q[m], 1))`.